File: frontend/src/lib/marketHours.ts
import { isUsEquityTradingDay } from "@/features/charting/tradingCalendar";

/** Get current time in US Eastern. */
function nowET(): Date {
  return new Date(
    new Date().toLocaleString("en-US", { timeZone: "America/New_York" }),
  );
}

/** Minutes since midnight for an ET Date. */
function minutesSinceMidnight(et: Date): number {
  return et.getHours() * 60 + et.getMinutes();
}

function isoDateFromET(et: Date): string {
  const y = et.getFullYear();
  const m = String(et.getMonth() + 1).padStart(2, "0");
  const d = String(et.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

function isTradingDayET(et: Date): boolean {
  return isUsEquityTradingDay(isoDateFromET(et));
}

const OPEN = 9 * 60 + 30; // 9:30 AM ET
const CLOSE = 16 * 60; // 4:00 PM ET
const CLOSE_BUF = 16 * 60 + 5; // 4:05 PM ET

/**
 * True during regular trading hours (9:30 AM - 4:00 PM ET) on trading days.
 */
export function isMarketOpen(): boolean {
  const et = nowET();
  if (!isTradingDayET(et)) return false;
  const m = minutesSinceMidnight(et);
  return m >= OPEN && m < CLOSE;
}

/**
 * True during market hours + 5 min buffer (9:30 AM - 4:05 PM ET) on trading days.
 * Use this to gate auto-refresh calls; the buffer captures final settlement data.
 */
export function isWithinTradingSession(): boolean {
  const et = nowET();
  if (!isTradingDayET(et)) return false;
  const m = minutesSinceMidnight(et);
  return m >= OPEN && m < CLOSE_BUF;
}

/**
 * True between market close (4:00 PM ET) and midnight on trading days.
 * Used to trigger a once-per-day post-close data update.
 */
export function isAfterClose(): boolean {
  const et = nowET();
  if (!isTradingDayET(et)) return false;
  const m = minutesSinceMidnight(et);
  return m >= CLOSE;
}

/**
 * Return today's date string in ET (YYYY-MM-DD).
 */
export function todayET(): string {
  return isoDateFromET(nowET());
}

File: backend/app/market_hours.py
"""US equity market hours utilities (Eastern Time)."""

from datetime import datetime, date, timedelta, time
import zoneinfo

ET = zoneinfo.ZoneInfo("America/New_York")

MARKET_OPEN = time(9, 30)
MARKET_CLOSE = time(16, 0)
MARKET_CLOSE_BUFFER = time(16, 5)  # +5 min buffer for data delays


def now_et() -> datetime:
    """Current datetime in US Eastern."""
    return datetime.now(ET)


def is_weekday(d: date) -> bool:
    return d.weekday() < 5


def is_market_open() -> bool:
    """True during regular trading hours (9:30 AM – 4:00 PM ET, weekdays)."""
    dt = now_et()
    if dt.weekday() >= 5:
        return False
    t = dt.time()
    return MARKET_OPEN <= t < MARKET_CLOSE


def is_within_trading_session() -> bool:
    """True during market hours + 5 min buffer (9:30 AM – 4:05 PM ET, weekdays).

    Use this to gate auto-refresh calls that should only run while the
    market is open or just after close (to capture final data).
    """
    dt = now_et()
    if dt.weekday() >= 5:
        return False
    t = dt.time()
    return MARKET_OPEN <= t < MARKET_CLOSE_BUFFER


def is_after_close() -> bool:
    """True between market close (4:00 PM ET) and next market open (9:30 AM ET).

    This is the window for post-close tasks like allocation snapshots.
    """
    dt = now_et()
    if dt.weekday() >= 5:
        return False  # weekends handled separately
    t = dt.time()
    return t >= MARKET_CLOSE or t < MARKET_OPEN


def get_allocation_target_date() -> date:
    """Determine the effective trading date for allocation snapshots.

    Rules:
    - Between 4:00 PM and midnight ET → next calendar day
    - Between midnight and 9:30 AM ET → same calendar day (today)
    - If the target falls on a weekend, roll forward to Monday.

    Returns None if called during market hours (allocations shouldn't
    be captured then).
    """
    dt = now_et()
    t = dt.time()
    today = dt.date()

    if t >= MARKET_CLOSE:
        # After close → target is next calendar day
        target = today + timedelta(days=1)
    elif t < MARKET_OPEN:
        # Before open → target is today
        target = today
    else:
        # During market hours — not the right time for allocations
        return None

    # Roll forward past weekends
    while target.weekday() >= 5:
        target += timedelta(days=1)

    return target


def next_trading_day(d: date = None) -> date:
    """Return the next trading day after the given date (skips weekends)."""
    if d is None:
        d = now_et().date()
    d = d + timedelta(days=1)
    while d.weekday() >= 5:
        d += timedelta(days=1)
    return d
